helix-core/src/main/java/org/apache/helix/controller/stages/BestPossibleStateCalcStage.java

```diff
@@ -347,36 +347,39 @@ private boolean computeSingleResourceBestPossibleState(ClusterEvent event,
       LogUtil.logError(logger, _eventId, "Error computing assignment for resource " + resourceName
           + ". no rebalancer found. rebalancer: " + rebalancer + " mappingCalculator: "
           + mappingCalculator);
-    }
-
-    if (rebalancer != null && mappingCalculator != null) {
-      ResourceAssignment partitionStateAssignment = null;
+    } else {
       try {
         HelixManager manager = event.getAttribute(AttributeName.helixmanager.name());
         rebalancer.init(manager);
         idealState =
             rebalancer.computeNewIdealState(resourceName, idealState, currentStateOutput, cache);
 
-        output.setPreferenceLists(resourceName, idealState.getPreferenceLists());
+        // Check if calculation is done successfully
+        if (!checkBestPossibleStateCalculation(idealState)) {
+          LogUtil.logWarn(logger, _eventId,
+              "The calculated idealState is not valid, resource: " + resourceName);
+          return false;
+        }
 
         // Use the internal MappingCalculator interface to compute the final assignment
         // The next release will support rebalancers that compute the mapping from start to finish
-        partitionStateAssignment = mappingCalculator
+        ResourceAssignment partitionStateAssignment = mappingCalculator
             .computeBestPossiblePartitionState(cache, idealState, resource, currentStateOutput);
 
         if (partitionStateAssignment == null) {
           LogUtil.logWarn(logger, _eventId,
-              "PartitionStateAssignment is null, resource: " + resourceName);
+              "The calculated partitionStateAssignment is null, resource: " + resourceName);
           return false;
         }
 
+        // Set the calculated result to the output.
+        output.setPreferenceLists(resourceName, idealState.getPreferenceLists());
```

Comment:
Is this to make sure if we ever get a null value for partitionStateAssignment, we don't set anything in the output? 

I wonder if this might cause a behavior change. Before this change, did `output` get populated with some preferenceList value (maybe the one from the previous, last successful pipeline)? After this change, if the `computeBestPossiblePartitionState` returns a `null`, then output will start having nothing, which is a different behavior from before this change?

Reply:
Yeah, this is a behavior change. But the original behavior is the problem here. If any value is set in the output, then the later stage will just use without any doubt. This is very dangerous.

I checked later stages like intermediate state calc, and message generating, they can handle a null output gracefully. But they cannot handle the output with the wrong information gracefully.=

Reply:
It seems to me that the original code, `return checkBestPossibleStateCalculation(idealState);` would still return false. Does it feed the wrong info in `output` to later stage? 

The main thing is that the new code fixed the issue of missing metric for invalid idealState.

Reply:
The original code returns false while the output is filled with an empty mapping. However, this empty mapping is misleading.